### PR TITLE
Webserver tweaks

### PR DESCRIPTION
**Pull request by phlptp**
### Summary

If merged this pull request will change the default webserver port to 8080.  Add some Environmental variable and command line support for changing the port on the webserver and some other broker servers.  And fix a few issues with the webserver and discrimination of target and queries.  As well as adding a main page to the server.

### CODE REVIEW

**Comment by phlptp** [comment]
I am not going to fix the clang-tidy issues for this PR.  Some of them have been fixed in HELICS 3 so don't want to do it again and some are issues that can't be fixed as they are in the boost code.

File: src/helics/apps/helics-broker-server.cpp [hunk 1]
```diff
@@ -74,8 +74,11 @@ int main(int argc, char* argv[])
             brokerServer->startServers();
             std::cout << "servers started\n";
             std::this_thread::sleep_for(opTime.to_ms());
+            std::cout
+                << "Helics broker server time limit reached servers closing (to change duration use `--duration=X`)"
```
Comment (nightlark) [inline src/helics/apps/helics-broker-server.cpp:78]:
```suggestion
                << "Helics broker server time limit reached, servers closing (to change duration use `--duration=X`)"
```

File: src/helics/apps/BrokerServer.cpp [hunk 2]
```diff
@@ -146,6 +167,12 @@ namespace apps {
         app->add_flag("--websocket",
                       websocket_server,
                       "start a websocket to respond to api requests");
+        app->add_option("--http_server_args",
+                        mHttpArgs,
+                        "command line arguments for the http server");
+        app->add_option("--websocket_server_args",
+                        mWebSocketArgs,
+                        "command line arguments for the web socket server");
```
Comment (nightlark) [inline src/helics/apps/BrokerServer.cpp:175]:
```suggestion
                        "command line arguments for the websocket server");
```

File: src/helics/apps/helicsWebServer.cpp [hunk 3]
```diff
@@ -870,6 +881,33 @@ static const Json::Value null;
...
+            ->envname("HELICS_HTTP_PORT");
+        parser.add_option("--http_interface",
+                          httpAddress_,
+                          "specify the interface to use for connecting an http server");
+
+        parser.add_option("--websocket_port", websocketPort_, "specify the websocket port to use")
+            ->envname("HELICS_WEBSOCKET_PORT");
+        parser.add_option("--websocket_interface",
+                          websocketAddress_,
+                          "specify the interface to use for connecting a we server");
```
Comment (nightlark) [inline src/helics/apps/helicsWebServer.cpp:900]:
```suggestion
                          "specify the interface to use for connecting a web server");
```

File: src/helics/apps/helicsWebServer.cpp [hunk 4]
```diff
@@ -870,6 +881,33 @@ static const Json::Value null;
...
+    void WebServer::processArgs(const std::string& args)
+
+    {
+        CLI::App parser("http web server parser");
+        parser.allow_extras();
+        parser.add_option("--http_port", httpPort_, "specify the http port to use")
+            ->envname("HELICS_HTTP_PORT");
+        parser.add_option("--http_interface",
+                          httpAddress_,
+                          "specify the interface to use for connecting an http server");
```
Comment (nightlark) [inline src/helics/apps/helicsWebServer.cpp:894]:
```suggestion
                          "specify the interface to use for connecting a http server");
```

File: src/helics/apps/helicsWebServer.cpp [hunk 5]
```diff
@@ -870,6 +881,33 @@ static const Json::Value null;
...
+            ->envname("HELICS_WEBSOCKET_PORT");
+        parser.add_option("--websocket_interface",
+                          websocketAddress_,
+                          "specify the interface to use for connecting a we server");
+
+        try {
+            parser.parse(args);
+        }
+        catch (const CLI::Error& ce) {
+            logMessage(std::string("error processing command line arguments for weberver :") +
```
Comment (nightlark) [inline src/helics/apps/helicsWebServer.cpp:906]:
```suggestion
            logMessage(std::string("error processing command line arguments for web server :") +
```

File: src/helics/apps/helicsWebServer.cpp [hunk 6]
```diff
@@ -765,7 +776,7 @@ class HttpSession: public std::enable_shared_from_this<HttpSession> {
         boost::ignore_unused(bytes_transferred);
 
         if (ec) {
-            return fail(ec, "write");
+            return fail(ec, "helics webserver write");
```
Comment (nightlark) [inline src/helics/apps/helicsWebServer.cpp:779]:
```suggestion
            return fail(ec, "helics web server write");
```

File: src/helics/apps/helicsWebServer.cpp [hunk 7]
```diff
@@ -751,7 +762,7 @@ class HttpSession: public std::enable_shared_from_this<HttpSession> {
 
         if (ec) {
             if (beast::error::timeout != ec) {
-                fail(ec, "read");
+                fail(ec, "helics webserver read");
```
Comment (nightlark) [inline src/helics/apps/helicsWebServer.cpp:765]:
```suggestion
                fail(ec, "helics web server read");
```

File: src/helics/apps/helicsWebServer.cpp [hunk 8]
```diff
@@ -510,7 +520,7 @@ class WebSocketsession: public std::enable_shared_from_this<WebSocketsession> {
         }
 
         if (ec) {
-            fail(ec, "read");
+            return fail(ec, "helics webserver read");
```
Comment (nightlark) [inline src/helics/apps/helicsWebServer.cpp:523]:
```suggestion
            return fail(ec, "helics web server read");
```

File: src/helics/apps/zmqBrokerServer.cpp [hunk 9]
```diff
@@ -26,6 +27,26 @@ static const Json::Value null;
...
+        parser.add_option("--zmq_port", mZmqPort, "specify the zmq port to use");
+        parser.add_option("--zmq_interface",
+                          mZmqInterface,
+                          "specify the interface to use for connecting the zmq broker server");
+
+        try {
+            parser.parse(args);
+        }
+        catch (const CLI::Error& ce) {
+            logMessage(std::string("error processing command line arguments for weberver :") +
```
Comment (nightlark) [inline src/helics/apps/zmqBrokerServer.cpp:45]:
```suggestion
            logMessage(std::string("error processing command line arguments for web server :") +
```

File: docs/user-guide/webserver.md [hunk 10]
```diff
@@ -35,10 +35,16 @@ The format is json.
 Then it can be specified on the command line like so:
 
 ```sh
-helics_broker_server --http --web --zmq --config broker_server_config.json
+helics_broker_server --web --zmq --config broker_server_config.json
 ```
 
-The configuration will then make the REST webserver accessible on any interface on port 8080 and a websocket server on port 8008.
+The configuration will then make the REST webserver accessible on any interface on port 8080 and a websocket server on port 8080. The port in use can be specified in a configuration file, or via command line such as
```
Comment (nightlark) [inline docs/user-guide/webserver.md:41]:
```suggestion
The configuration will then make the REST web server accessible on any interface on port 8080 and a WebSocket server on port 8080. The port in use can be specified in a configuration file, or via command line such as
```

File: docs/user-guide/webserver.md [hunk 11]
```diff
@@ -35,10 +35,16 @@ The format is json.
...
+
+```sh
+helics_broker_server --web --zmq --http_server_args="--http_port=80"
+```
+
+valid arguments are of the form `--<server>_server_args` in that arg field `--<server>_port` and `--<server>_interface`. valid server names are `http`, `websocket`, `zmq`, 'tcp', and `udp` and eventually `mpi`. The http web server also acknowledges `HELICS_HTTP_PORT` as an environmental value. The websocket server acknowledges `HELICS_WEBSOCKET_PORT` for the port numbers of the respective servers.
```
Comment (nightlark) [inline docs/user-guide/webserver.md:47]:
```suggestion
Arguments are passed to servers using an option in the form `--<server>_server_args`, and in that arg field `--<server>_port` and `--<server>_interface` are valid arguments. Valid server names are `http`, `websocket`, `zmq`, 'tcp', and `udp`, and eventually `mpi`. The http web server also acknowledges `HELICS_HTTP_PORT` as an environment variable. The websocket server acknowledges `HELICS_WEBSOCKET_PORT` for the port numbers of the respective servers.
```

File: src/helics/apps/indexPage.hpp [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+#pragma once
+
+namespace helics {
+namespace webserver {
+
+    static const char* indexPage1 =
+        R"raw(<html>
+  <head>
+    <meta charset="utf-8" />
+    <title>HELICS Web server interface</title>
```
Comment (nightlark) [inline src/helics/apps/indexPage.hpp:16]:
```suggestion
    <title>HELICS web server interface</title>
```

**Review by nightlark — COMMENTED** [review]
A few doc/string suggestions.

File: src/helics/apps/helicsWebServer.cpp (hunk 4 above)
Comment (phlptp) [reply src/helics/apps/helicsWebServer.cpp:896]: I think this is supposed to be `an` since you are saying the letter which starts with a vowel sound

Comment (nightlark) [reply src/helics/apps/helicsWebServer.cpp:896]: Yea, I think that's right for the most common pronunciation.

**Review by kdheepak — APPROVED** [review]

**Review by nightlark — APPROVED** [review]

**Pull request merged**